contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,853 @@
+the StatefulSet controller will do the following.
+    1. The controller will 
+    [reconstruct the revision history](#history-reconsturction) of the 
+    StatefulSet.
+    1. If the revision history of the StatefulSet contains a PodTemplate 
+    whose `.PodTemplateSpec` is semantically, deeply equivalent to the 
+    StatefulSet's `.Spec.Template`, the youngest such PodTemplate will be used 
+    as the target PodTemplate. 
+    1. If no such PodTemplate exists, the StatefulSet controller will 
+    [create a new PodTemplate](#podtemplate-creation) from the StatefulSet's 
+    `.Spec.Template`, and it will use this as the target PodTemplate.
+    1. The controller will update the StatefulSet's `.Status.TargetTemplate` 
+    based on the selection made above.
+
+### PodTemplate Selection
+When the StatefulSet controller creates the Pods in a StatefulSet, it will use 
+the following criteria to select the PodTemplateSpec used to create a 
+Pod. These criteria allow the controller to continue to make progress toward 
+its target state, while respecting its guarantees and allowing for rolling 
+updates back and forward.
+
+1. If the StatefulSet's `.Spec.TemplateGeneration` is nil, then the cluster 
+has been upgraded from a version that does not support StatefulSet update to 
+a version that does. 
+    1. In this case the `.Spec.Template` is the current revision, 
+    and no Pods in the StatefulSet should be labeled with a 
+    `StatefulSetPodTemplateGeneration` label. 
+    1. The StatefulSet will initialize its revision history on the first 
+    update to its `.Spec.Template`.
+1. If the StatefulSet's `.Spec.TemplateGeneration` is equal to its 
+`.Status.TemplateGeneration`, then there is no update in progress and all 
+Pods will be created from the PodTemplate matching this revision.
+1. If the Pod's ordinal is in the sequence `[0,.Status.CurrentReplicas)`, 
+then it was previously created from the PodTemplate matching the 
+StatefulSet's `.Status.TemplateGeneration`, and it will be recreated 
+from this PodTemplate.
+1. If the Pod's ordinal is in the sequence
+ `[.Spec.Replicas-.Status.UpdatedReplicas,.Spec.Replicas)`, then it was
+ previously created from the PodTemplate matching the StatefulSet's, 
+ `.Status.TargetTemplateGeneration`, and it will be recreated from this 
+ PodTemplate.
+1. If the ordinal does not meet either of the prior two conditions, and 
+if ordinal is in the sequence `[0, .Spec.GenerationPartition)`, it will be created 
+from the PodTemplate matching the StatefulSet's 
+`.Status.TemplateGeneration`.
+1. Otherwise, the Pod is created from the PodTemplate matching the 
+StatefulSet's `.Status.TargetTemplateGeneration`. 
+
+### Update Completion
+A StatefulSet update is complete when the following conditions are met.
+
+1. All Pods with ordinals in the sequence `[0,.Spec.Replicas)` have a Status of 
+Running and a Ready Condition.
+1. The StatefulSet's `.Spec.GenerationPartition` is equal to `0`. 
```

Comment:
@kow3ns can you make sure that @bgrant0607 is also looped in because he had some thoughts about canaries too (see https://github.com/kubernetes/kubernetes/issues/11505).